src/skvhttpclient/src/skvhttp/client/SKVClient.cpp

```diff
@@ -0,0 +1,157 @@
+}
+
+boost::future<Response<std::shared_ptr<dto::Schema>>> Client::getSchema(const String& collectionName, const String& schemaName, int64_t schemaVersion) {
+  SchemaCacheT::const_iterator collectionIt = _schemaCache.find(collectionName);
+  if (collectionIt != _schemaCache.end()) {
+    auto schemaNameIt = collectionIt->second.find(schemaName);
+    if (schemaNameIt != collectionIt->second.end()) {
+      if (schemaVersion == dto::ANY_SCHEMA_VERSION && schemaNameIt->second.size()) {
+        std::shared_ptr<dto::Schema> schema = schemaNameIt->second.begin()->second;
+        return MakeResponse<std::shared_ptr<dto::Schema>>(Statuses::S200_OK("Got schema from cache"), std::move(schema));
+      }
+      
+      auto schemaVersionIt = schemaNameIt->second.find(schemaVersion);
+      if (schemaVersionIt != schemaNameIt->second.end()) {
+        std::shared_ptr<dto::Schema> schema = schemaVersionIt->second;
+        return MakeResponse<std::shared_ptr<dto::Schema>>(Statuses::S200_OK("Got schema from cache"), std::move(schema));
+      }
+    }
+  }
+  
+  dto::GetSchemaRequest request {
+    .collectionName = collectionName,
+    .schemaName = schemaName,
+    .schemaVersion = schemaVersion
+  };
+
+  return _HTTPClient.POST<dto::GetSchemaRequest, dto::GetSchemaResponse>("/api/GetSchema", std::move(request)).
+    then([this, collectionName, schemaName, schemaVersion] (auto&& futResp) {
+      auto&& [status, resp] = futResp.get();
+      auto schema = std::make_shared<dto::Schema>(resp.schema);
+      if (status.is2xxOK()) {
+        _schemaCache[collectionName][schemaName][schemaVersion] = schema;
+      }
+      return Response<std::shared_ptr<dto::Schema>>(std::move(status), schema);
+    });
+}
+
+boost::future<Response<TxnHandle>> Client::beginTxn(dto::TxnOptions options) {
+  dto::TxnBeginRequest request {
+    .options = std::move(options)
+  };
+
+   return _HTTPClient.POST<dto::TxnBeginRequest, dto::TxnBeginResponse>("/api/TxnBegin", std::move(request)).
+    then([this] (auto&& futResp) {
+      auto&& [status, resp] = futResp.get();
+      TxnHandle txn(this, resp.timestamp);
+      return Response<TxnHandle>(std::move(status), std::move(txn));
+    });
+}
+
+boost::future<Response<>> TxnHandle::write(dto::SKVRecord& record, bool erase, dto::ExistencePrecondition precondition) {
+  dto::WriteRequest request {
+    .timestamp = _id,
+    .collectionName = record.collectionName,
+    .schemaName = record.schema->name,
+    .isDelete = erase,
+    .precondition = precondition,
+    .value = record.storage.share(),
+    .fieldsForPartialUpdate = std::vector<uint32_t>{}
+  };
+
+  return _client->_HTTPClient.POST<dto::WriteRequest>("/api/Write", std::move(request));
+}
+
+boost::future<Response<dto::SKVRecord>> TxnHandle::read(dto::SKVRecord record) {
+  if (record.fieldCursor != record.schema->fields.size()) {
+    return MakeResponse(Statuses::S400_Bad_Request("All fields of record must be serialized for read request"), dto::SKVRecord{});
```

Comment:
really? I thought only keys are needed

Reply:
The problem is the proxy tries to deserialize all the fields to make the storage object on its side, and we can't fix it here for the user since we need the builder for that. We could take a builder as the parameter here but then the user can't reuse records or use reconstructed records like for secondary index usage.

Reply:
i see. i think it's fine to leave as-is